### PR TITLE
mesh: add validation for the new pbmesh resources

### PR DESCRIPTION
**Pull request by rboyer**
### Description

Adds validation for HTTPRoute, GRPCRoute, TCPRoute, DestinationPolicy, and ComputedRoutes.

NET-5054
NET-5055

### CODE REVIEW

File: internal/mesh/internal/types/grpc_route_test.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	"github.com/stretchr/testify/require"
+
+	"github.com/hashicorp/consul/internal/catalog"
+	"github.com/hashicorp/consul/internal/resource/resourcetest"
+	pbmesh "github.com/hashicorp/consul/proto-public/pbmesh/v1alpha1"
+	"github.com/hashicorp/consul/proto/private/prototest"
+	"github.com/hashicorp/consul/sdk/testutil"
+)
+
+func TestValidateGRPCRoute(t *testing.T) {
```
Comment (ishustava) [inline internal/mesh/internal/types/grpc_route_test.go:18]: I ran these tests with coverage and I think there's some missing test cases that could be covered.

File: internal/mesh/internal/types/http_route.go [hunk 2]
```diff
@@ -27,6 +35,546 @@ func RegisterHTTPRoute(r resource.Registry) {
 	r.Register(resource.Registration{
 		Type:     HTTPRouteV1Alpha1Type,
 		Proto:    &pbmesh.HTTPRoute{},
-		Validate: nil,
+		Mutate:   MutateHTTPRoute,
+		Validate: ValidateHTTPRoute,
 	})
 }
+
+func MutateHTTPRoute(res *pbresource.Resource) error {
```
Comment (ishustava) [inline internal/mesh/internal/types/http_route.go:43]: I think this one needs tests

File: internal/mesh/internal/types/http_route_test.go [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+		prototest.AssertDeepEqual(t, tc.route, got.Data)
+
+		if tc.expectErr == "" {
+			require.NoError(t, err)
+		} else {
+			testutil.RequireErrorContains(t, err, tc.expectErr)
+		}
+	}
+
+	cases := map[string]testcase{
```
Comment (ishustava) [inline internal/mesh/internal/types/http_route_test.go:45]: There's probably some missing cases here and I see the todos below. Did you want to add them later?

**Review by ishustava — COMMENTED** [review]
Looks great! I mostly left some comments about missing tests but otherwise it looks good to me.

File: internal/mesh/internal/types/computed_routes.go [hunk 4]
```diff
@@ -27,6 +29,49 @@ func RegisterComputedRoutes(r resource.Registry) {
...
+	var merr error
+
+	if len(config.PortedConfigs) == 0 {
+		merr = multierror.Append(merr, resource.ErrInvalidField{
+			Name:    "ported_configs",
+			Wrapped: resource.ErrEmpty,
+		})
+	}
+
+	// TODO(rb): do more elaborate validation
```
[the remark this answers is not in the document]
Comment (rboyer) [reply internal/mesh/internal/types/computed_routes.go:52]: I was going to come back to this in NET-5066 when we can update all of the validation to handle tenancy normalization and tenancy tests for inner references.

File: internal/mesh/internal/types/http_route_test.go (hunk 3 above)
Comment (rboyer) [reply internal/mesh/internal/types/http_route_test.go:169]: That was what NET-5066 was for.

File: internal/mesh/internal/types/grpc_route_test.go (hunk 1 above)
Comment (rboyer) [reply internal/mesh/internal/types/grpc_route_test.go:18]: That was what NET-5066 was for.

**Comment by rboyer** [comment]
The validation/mutation hooks for the xRoute types has had the code coverage increased.

File: internal/mesh/internal/types/http_route.go [hunk 5]
```diff
@@ -27,6 +35,546 @@ func RegisterHTTPRoute(r resource.Registry) {
...
+	if err := validateParentRefs(route.ParentRefs); err != nil {
+		merr = multierror.Append(merr, err)
+	}
+
+	if len(route.Hostnames) > 0 {
+		merr = multierror.Append(merr, resource.ErrInvalidField{
+			Name:    "hostnames",
+			Wrapped: errors.New("should not populate hostnames"),
+		})
+	}
```
[the remark this answers is not in the document]
Comment (rboyer) [reply internal/mesh/internal/types/http_route.go:90]: Yes, but they'll just come back in when api GW stuff migrates.

File: internal/mesh/internal/types/grpc_route.go [hunk 6]
```diff
@@ -25,8 +29,183 @@ var (
...
+		for j, match := range rule.Matches {
+			wrapMatchErr := func(err error) error {
+				return wrapRuleErr(resource.ErrInvalidListElement{
+					Name:    "matches",
+					Index:   j,
+					Wrapped: err,
+				})
+			}
+
+			if match.Method != nil && match.Method.Type == pbmesh.GRPCMethodMatchType_GRPC_METHOD_MATCH_TYPE_UNSPECIFIED {
```
[the remark this answers is not in the document]
Comment (rboyer) [reply internal/mesh/internal/types/grpc_route.go]: https://protobuf.dev/programming-guides/enum/

I did not realize that they changed how enum downgrade works from proto2 to proto3. Checking for unset/unspecified/default was the right way to handle out-of-spec future values in older code when they were _closed_ enums, but I guess nowadays they're not. I have to go stare at a bunch of code now and patch it.

[the remark this answers is not in the document]
Comment (rboyer) [reply internal/mesh/internal/types/grpc_route.go]: Fixed now, and tested for HTTPRoute, GRPCRoute, DestinationPolicy. Opened a different PR to cover FailoverPolicy.

File: internal/mesh/internal/types/http_route.go [hunk 7]
```diff
@@ -27,6 +35,546 @@ func RegisterHTTPRoute(r resource.Registry) {
...
+					},
+				))
+			}
+		}
+	}
+
+	return merr
+}
+
+func validateHTTPTimeouts(timeouts *pbmesh.HTTPRouteTimeouts) []error {
```
[the remark this answers is not in the document]
Comment (rboyer) [reply internal/mesh/internal/types/http_route.go]: Moved things that can be shared into xroute.go.

**Review by mkeeler — APPROVED** [review]
Approving to unblock subsequent work. We can follow on with any extra validations later closer to when the features will become GA.

**Pull request merged**